test/construct.test.did

```diff
@@ -85,14 +85,33 @@ assert blob "DIDL\01\6c\02\00\7c\01\7e\02\7e\01\00\2a\01\00"                   !
 assert blob "DIDL\01\6c\02\00\01\01\7e\01\7c\2a"                               !: (reserved) "record: type out of range";
 
+// opt
+assert blob "DIDL\00\01\7f" == "(null)"                                         : (opt empty) "opt: parsing null : null";
+assert blob "DIDL\01\6e\6f\01\00\00" == "(null)"                                : (opt empty) "opt: parsing null : opt empty";
+assert blob "DIDL\01\6e\7e\01\00\00" == "(null)"                                : (opt bool) "opt: parsing null : opt bool ";
+assert blob "DIDL\01\6e\7e\01\00\01\00" == "(opt true)"                         : (opt bool) "opt: parsing opt true : opt bool";
+assert blob "DIDL\01\6e\7e\01\00\01\01" == "(opt false)"                        : (opt bool) "opt: parsing opt false : opt bool";
+assert blob "DIDL\01\6e\7e\01\00\01\02"                                        !: (opt bool) "opt: parsing invalid bool at opt bool";
+
+// opt subtype to constituent
+assert blob "DIDL\00\01\7e\00" == "(opt true)"                                  : (opt bool) "opt: parsing opt true : opt bool";
+assert blob "DIDL\00\01\7e\01" == "(opt false)"                                 : (opt bool) "opt: parsing opt false : opt bool";
+assert blob "DIDL\00\01\7e\02"                                                 !: (opt bool) "opt: parsing invalid bool at opt bool";
+assert blob "DIDL\01\6e\7f\01\00\00" == "(null)"                                : (opt opt null) "opt: parsing (null : opt null) at opt opt null gives null, not opt null";
+
+// special opt and record subtyping
+assert blob "DIDL\01\6e\7e\01\00\00"    == "(null)"  : (opt nat) "null : opt bool <: opt nat";
+assert blob "DIDL\01\6e\7e\01\00\01\01" == "(null)"  : (opt nat) "opt true : opt bool <: opt nat";
+assert blob "DIDL\01\6e\7e\01\00\01\02"             !: (opt nat) "opt bool <: opt nat with invalid boolean value";
+assert blob "DIDL\01\6e\6e\7e\01\00\01\01\01" == "(null)"  : (opt nat) "opt opt true : opt opt bool <: opt nat";
+assert blob "DIDL\01\6e\6e\7e\01\00\01\01\01" == "(opt null)"  : (opt opt nat) "opt opt true : opt opt bool <: opt opt nat";
+
 // variant
 assert "(variant {})"                                                          !: (variant {}) "variant: no empty value";
 assert blob "DIDL\01\6b\00\01\00"                                              !: (variant {}) "variant: no empty value";
 assert blob "DIDL\01\6b\01\00\7f\01\00\00" == "(variant {0})"                   : (variant {0}) "variant: numbered field";
 assert blob "DIDL\01\6b\01\00\7f\01\00\00\2a"                                  !: (variant {0:int}) "variant: type mismatch";
-
-// TODO Should this pass?
-assert blob "DIDL\01\6b\02\00\7f\01\7c\01\00\01\2a"                             : (variant {0:int; 1:int}) "??? variant: type mismatch";
+assert blob "DIDL\01\6b\02\00\7f\01\7c\01\00\01\2a"                             : (variant {0:int; 1:int}) "variant: type mismatch in unused tag";
```

Comment:
Oh, I thought that clarification had landed (the formula with the ∃t.) but it was only in the PR discussion maybe?  Ah, no, it’s here:

> The effect of this rule is that decoders do not actually need to perform separate subtype checks during deserialisation. Instead, they can simply try to deserialise, and if they encounter a mismatch abort to the innermost option type, returning null.
> 
> Note: The working assumption is that the type describing an incoming value ought to be principal, i.e., the most precise type assignable to the value (in a sense that could be made precise). In that case, <datatype''> equals <datatype> and deserialization succeeds exactly if the types match.
> 
> However, in practice it would be costly to enforce the requirement that all type descriptions in a serialised value are principal, for both encoder (who would need to compute the principal type) and decoders (who would need to check it). For example, to be principal, an empty vector would need to have type vec empty, null could only have type null, and a variant would only be allowed to include tags that actually occur in the value. Consequently, the semantics allows for the possibility that the encoder ascribes a less specific type <datatype> with some redundant information, and allows the decoder to ignore redundant information (by going to <datatype''>) like the element type of empty arrays, of null options, or unused variant tags.

I am really inclined to rewrite the  (kinda organically grown) spec so that all this becomes clear, and it isn’t hidden in prose.

Reply:
Gave it a shot at https://github.com/dfinity/candid/pull/128